### DISCUSSION TITLE
ios opengles sample texture error （use stb_image）while android is ok

### DISCUSSION THREAD

**Discussion by zhangbaochong** [nothings/stb#1313] [Q&A] [answered]
https://stackoverflow.com/questions/71566579/ios-opengles-sample-texture-error-use-stb-image-while-android-is-ok

I use stb_image load texture, such as

`stbi_set_flip_vertically_on_load(true);
GLuint textureID;
glGenTextures(1, &textureID);
int width, height, nrComponents;
unsigned char *data = stbi_load(path, &width, &height, &nrComponents, 0);

GLenum format;
if (nrComponents == 1) {
    format = GL_RED;
} else if (nrComponents == 3) {
    format = GL_RGB;
} else {
    format = GL_RGBA;
}

glTexImage2D(GL_TEXTURE_2D, 0, format, width, height, 0, format, GL_UNSIGNED_BYTE, data);
glGenerateMipmap(GL_TEXTURE_2D);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);`

The same code in android is ok, while in ios the rendering result get some problem:
<img width="360" alt="20220322-112238(WeLinkPC)" src="https://user-images.githubusercontent.com/16303584/159437271-6bae9222-cc77-4712-9d87-a848dda87fc2.png">
if set stbi_set_flip_vertically_on_load(false):
<img width="334" alt="2" src="https://user-images.githubusercontent.com/16303584/159437350-72895e3a-a906-4059-a8ee-48f98ded3ea3.png">
the correct result with android:
![1](https://user-images.githubusercontent.com/16303584/159437436-792adcef-1128-49ea-926b-cab279c59dba.png)

**Comment by nothings** [comment]
the last image does not look like it's the same image as the first two. are you sure it is?

**Reply by zhangbaochong** [reply]
its just different camera angle.

**Reply by zhangbaochong** [reply]
[model.zip](https://github.com/nothings/stb/files/8322732/model.zip)

**Reply by nothings** [reply]
no idea

**Reply by zhangbaochong** [reply]
its ok if I sample texture to a square, so it may be not an issues with stb_image. thanks

**Reply by mmozeiko** [reply]
I don't know on which GL version iOS is, but in OpenGL ES 2.0 non-power-of-two textures support only `GL_CLAMP_TO_EDGE` wrap mode and `GL_NEAREST` or `GL_LINEAR` filter modes. Other combinations are not supported unless special extensions are present.

**Reply by zhangbaochong** [reply]
Its OpenGLES 3.0 and my texutre also is power-of-two texture.

**Reply by nothings** [reply]
The texture inside model.zip is 1280x1280, which is not a power-of-two.

**Comment by zhangbaochong** [comment, marked as answer]
The issue is that GLKView not set depth buffer, just add one line code : `view.drawableDepthFormat = GLKViewDrawableDepthFormat24;`
